### ISSUE TITLE
docs(train): document preference training (DPO/ORPO/KTO) + updated response masking from PR #4358

### ISSUE THREAD

**Issue by MervinPraison** [MervinPraison/PraisonAIDocs#3342]
## Context

Source PR: [MervinPraison/PraisonAI#4358 — fix(train): actually mask prompts out of the loss](https://github.com/MervinPraison/PraisonAI/pull/4358) (merged 2026-08-25).

The PR is a large training-package update. It fixes a silent correctness bug in loss masking, adds three new training methods (DPO, ORPO, KTO) that were previously unreachable, and adds several new config keys. The docs currently describe **only SFT** — DPO/ORPO/KTO have **zero coverage** in `MervinPraison/PraisonAIDocs` (verified via code search). The masking behaviour documented in `docs/train.mdx` is now materially out of date.

Source of truth for every change below: `src/praisonai-train/praisonai_train/train/llm/trainer.py` at head `2be4685f1bb2e7e0d47a5b233ca2a45e0afc2db8`.

Follow `AGENTS.md`:
- Read the SDK source first, don't guess.
- Place new pages in `docs/features/` (not `docs/concepts/`).
- User-facing, agent-centric examples first; progressive disclosure.
- Hero Mermaid diagram, `Steps` / `AccordionGroup` / `CardGroup`, standard palette.
- Do NOT edit `docs/sdk/reference/**` — those are auto-generated.

---

### 1. Preference-tuning support — NEW (undocumented today)

A `method` config key selects the trainer. Registered in `TRAINING_METHODS` in `trainer.py`:

| method | Trainer | Config class | Required dataset columns | Needs ref model? | Summary |
|---|---|---|---|---|---|
| `sft` (default) | `SFTTrainer` | `SFTConfig` | — (uses `text` column) | No | supervised fine-tuning on completions |
| `dpo` | `DPOTrainer` | `DPOConfig` | `prompt`, `chosen`, `rejected` | Yes (auto-uses frozen base weights via `ref_model=None`) | direct preference optimisation on chosen/rejected pairs |
| `orpo` | `ORPOTrainer` | `ORPOConfig` | `prompt`, `chosen`, `rejected` | No (ORPO folds the reference model into its loss) | odds-ratio preference optimisation, no reference model |
| `kto` | `KTOTrainer` | `KTOConfig` | `prompt`, `completion`, `label` | Yes | Kahneman-Tversky optimisation on thumbs-up/down labels |

Behaviour details verified in source:
- `method` is normalised to lowercase and validated in `TrainModel.resolve_method()`; an unknown value raises `ValueError` listing every supported method with its summary.
- `TrainModel._require_columns()` fails **before** the run starts if the dataset does not have the required columns, and the error names both the missing columns and the columns the dataset actually has (TRL's own error surfaces deep in the collator, minutes into the run).
- `process_dataset()` short-circuits and returns the dataset **unmodified** when `method != "sft"` — preference datasets keep `prompt` / `chosen` / `rejected` (or `prompt` / `completion` / `label`) as-is; SFT keeps the existing flatten-to-`text` behaviour. (Shuffle still applies when `dataset[].shuffle: true`.)
- Preference imports are lazy and individually guarded: an older TRL missing one of `DPOTrainer` / `DPOConfig` / `ORPOTrainer` / `ORPOConfig` / `KTOTrainer` / `KTOConfig` still lets the other methods run.
- For preference methods the trainer drops SFT-only fields (`dataset_text_field`, `packing`, `dataset_num_proc`), sets `max_length = max_seq_length`, defaults `max_prompt_length` to `max_seq_length // 2`, and passes `beta` / `desirable_weight` / `undesirable_weight` when set.
- If the required trainer / config class is not importable from TRL, the trainer raises: `method 'X' needs YTrainer and YConfig from TRL, which this TRL version does not provide. Upgrade trl, or use method: sft.`

New config keys (must be added to the config reference table so the "unknown key" warning does not fire):

| Key | Type | Default | Applies to | Description |
|---|---|---|---|---|
| `method` | `str` | `"sft"` | all | Training method. One of `sft`, `dpo`, `orpo`, `kto`. Case-insensitive. |
| `beta` | `float` | trainer default | dpo, orpo, kto | Preference-loss temperature. Passed straight to the trainer's `beta`. |
| `max_prompt_length` | `int` | `max_seq_length // 2` | dpo, orpo, kto | Prompt truncation length for preference methods. |
| `desirable_weight` | `float` | trainer default | kto | Weight on thumbs-up examples. |
| `undesirable_weight` | `float` | trainer default | kto | Weight on thumbs-down examples. |

### 2. Response-only loss — behaviour materially changed (docs are stale)

**The old docs are wrong now.** They say `assistant_only_loss: auto` "enables masking only when the template supports it" and otherwise trains full-sequence. In reality **none of unsloth's 43 chat templates ship `{% generation %}` markers**, so on Gemma / Qwen / Llama / Mistral / Phi templates the "auto" default was silently training on prompts as well as answers. The PR fixes this by adding a second masking route.

Two routes now, chosen automatically:

1. `assistant_only_loss` (TRL's built-in) — used when the chat template has `{% generation %}` markers.
2. `train_on_responses_only` (Unsloth's `unsloth.chat_templates.train_on_responses_only`) — fallback that masks by locating literal turn-marker strings; works on any template family listed below.

Route selection is in `decide_masking(use_mask, supports_mask, markers)`:

| `use_mask` | `supports_mask` | `markers` known | Route |
|---|---|---|---|
| False | any | any | no masking (full sequence) |
| True | True | any | `assistant_only_loss` |
| True | False | yes | `train_on_responses_only` |
| True | False | no | `ValueError` — masking asked for but impossible |

`auto` resolves to `use_mask = supports_mask OR bool(markers)` — i.e. mask whenever any route is available.

Template → marker family table (from `TEMPLATE_TO_MARKERS` and `RESPONSE_MARKERS`):

| Chat template | Family | Instruction marker | Response marker |
|---|---|---|---|
| `llama-3`, `llama3`, `llama-3.1`, `llama-31`, `llama` | llama-3 | `<\|start_header_id\|>user<\|end_header_id\|>\n\n` | `<\|start_header_id\|>assistant<\|end_header_id\|>\n\n` |
| `chatml`, `qwen-2.5`, `qwen25`, `qwen2.5`, `qwen-25`, `qwen-3`, `qwen3`, `qwen3-instruct`, `qwen3-thinking`, `phi-4`, `gptoss`, `gpt-oss`, `yi-chat`, `unsloth`, `gemma_chatml`, `gemma2_chatml` | chatml | `<\|im_start\|>user\n` | `<\|im_start\|>assistant\n` |
| `gemma`, `gemma2`, `gemma-3`, `gemma3`, `gemma-3n`, `gemma3n`, `gemma-4`, `gemma4` | gemma | `<start_of_turn>user\n` | `<start_of_turn>model\n` |
| `mistral` | mistral | `[INST]` | `[/INST]` |
| `phi-3`, `phi-35`, `phi-3.5` | phi | `<\|user\|>\n` | `<\|assistant\|>\n` |
| `zephyr` | zephyr | `<\|user\|>\n` | `<\|assistant\|>\n` |

Resolution behaviour worth surfacing in docs:
- `resolve_response_markers(chat_template, model_name)` matches longest-first (so `llama-3.1` beats `llama`), is case-insensitive, and normalises hyphens/dots (`gemma-2` == `gemma2`).
- When `chat_template` is not set, the resolver falls back to the model name — but matches the **repo name only**, not the org, because every unsloth model is `unsloth/<name>` and `unsloth` is itself a template key (this bug is called out explicitly in the PR body).
- Unknown model → `None`. The trainer refuses to guess: masking on the wrong markers finds no boundary and trains on nothing, which is worse than not masking at all.

Updated `ValueError` text when `assistant_only_loss: true` but no route is available (replaces the older `{% generation %}`-only message currently quoted in `docs/train.mdx`):

```
assistant_only_loss is enabled but neither masking route is available for
'<model_name>': the chat template has no {% generation %} markers, and no
turn markers are known for it. Set assistant_only_loss: false to train on
the full sequence, or choose a chat_template from: <sorted list of known
template keys>.
```

Updated run-summary line — the "Loss mask" line now names which route ran (source: `_MASK_LABELS` in `trainer.py`):

| Route | Line printed |
|---|---|
| off | `Loss mask:   full sequence (training on prompts too)` |
| `assistant_only_loss` | `Loss mask:   assistant replies only (TRL assistant_only_loss)` |
| `train_on_responses_only` | `Loss mask:   assistant replies only (unsloth turn markers)` |

### 3. Small quality-of-life fixes

- **Single-element `quantization_method` list accepted.** Configs shipped by earlier templates write it as a single-element list; the trainer now unwraps `[x]` to `x`. Applied in both `for_export()` and `validate_config()`.
- **`dataset.processing_func` now warns.** The key was advertised in shipped templates but never read; it now logs a `WARNING: dataset.processing_func (...) is not supported and will be ignored; formatting is chosen automatically from the dataset's columns.`
- **CodeRabbit summary also mentions a DeepSeek stop-token export fix.** I did not see it in the primary diff — a doc writer should confirm from a follow-up commit or the ollama export path before including it. If it landed elsewhere, mention it in the Ollama export section; otherwise drop it.

---

### A. Update `docs/train.mdx`

1. **`### Assistant-only Loss` section (currently around line 447):**
   - Replace the "enables masking only when the template supports it" wording — it is no longer accurate.
   - Explain the two routes (`assistant_only_loss` vs `train_on_responses_only`) and the auto-selection rule.
   - Add a small Mermaid decision diagram (see palette in `AGENTS.md`).
   - Update the `Warning` block with the new `ValueError` text above.
   - Add the template → family table (or link to it in a new page — see B).
   - Update the run-summary line examples so users can recognise which route actually ran.

2. **`## Config.yaml example` (currently around line 671):**
   - Add `method` (defaulted to `sft`) with a comment on the four choices.
   - Add commented-out `beta`, `max_prompt_length`, `desirable_weight`, `undesirable_weight`.

3. **`### New / newly-honored fine-tuning config keys` (currently around line 846):**
   - Add rows for `method`, `beta`, `max_prompt_length`, `desirable_weight`, `undesirable_weight`.
   - Add a note that `quantization_method` accepts a single-element list.

4. **`### Silent-behaviour notes` / `## Silent-behaviour NOTES` (currently around lines 830 and 1280):**
   - Note the `dataset.processing_func` warning.
   - Note the single-element `quantization_method` list unwrap.

5. **`### Model-agnostic response-only training` (around line 1307) and `### Which chat_template should I use?` decision diagram (around line 1335):**
   - Rewrite: the marker-based route is now the answer for every template without `{% generation %}` (which is all of them today). Update the diagram nodes so the "Need assistant_only_loss?" branch no longer implies you must "Pick a template with generation blocks" — the marker fallback covers Gemma / Qwen / Llama / Mistral / Phi.
   - Remove the sentence "This uses TRL's model-agnostic `SFTConfig(assistant_only_loss=True)` instead of the old Llama-only `train_on_responses_only` markers." It's no longer accurate.

### B. New page — `docs/features/train-preference-tuning.mdx`

Preference tuning is a new user-facing capability; give it its own page rather than burying it in `docs/train.mdx`. Follow the AGENTS.md template exactly (frontmatter, hero Mermaid, `Steps`, config table, `AccordionGroup`, `CardGroup`).

Suggested outline:

- **Frontmatter**: `title: "Preference Training"`, `icon: "scale-balanced"` (or `"sliders"`), one-line description.
- **Hero Mermaid**: three lanes — SFT (single completions) vs DPO/ORPO (chosen vs rejected) vs KTO (thumbs-up/down).
- **Quick Start**: minimal `praisonai train` config per method — dataset shape + the two-or-three lines that switch from SFT.
  ```yaml
  method: dpo
  model_name: "unsloth/gemma-2-2b-it-bnb-4bit"
  max_seq_length: 2048
  max_prompt_length: 1024
  beta: 0.1
  dataset:
    - name: "your-org/your-preference-dataset"
      split: "train"
  ```
- **Which method should I use?** — a decision Mermaid diagram (SFT for behaviour-cloning, DPO/ORPO for pairwise preferences, KTO when you only have per-example labels; ORPO if VRAM-constrained).
- **Dataset shapes** — three small tables showing the exact columns and one example row per method.
- **Config keys**: `method`, `beta`, `max_prompt_length`, `desirable_weight`, `undesirable_weight`.
- **What happens under the hood**: masking is skipped (preference losses don't use it), the dataset is not flattened to `text`, SFT-only fields are dropped from the config, and DPO / KTO get `ref_model=None` (uses frozen base weights — works cleanly with a PEFT adapter and avoids doubling VRAM).
- **Errors you might see** and how to fix them: missing preference columns; `method '...' needs YTrainer and YConfig from TRL`; unknown method name.
- **Related** cards → `docs/train.mdx`, response-masking section.

### C. Update `docs/cli/train.mdx`

- Add `method`, `beta`, `max_prompt_length`, `desirable_weight`, `undesirable_weight` to the config table.
- Update the `assistant_only_loss` row description — remove "fails fast if unsupported" (only true when `true` **and** no marker route either).

### D. Update `docs/features/praisonai-train-package.mdx`

- The "Three headline safety guarantees" block currently says: `assistant_only_loss: auto` never crashes on a stock template — it falls back to full-sequence SFT. **This is factually out of date** and should say "it now masks via unsloth turn markers when the template lacks `{% generation %}`, and falls back to full-sequence SFT only when neither route is available."
- Add a one-liner that DPO / ORPO / KTO are supported alongside SFT, linking to the new preference-training page.

### E. `docs.json` navigation

- Add the new preference-training page under the **Features** group (not Concepts).
- Leave TypeScript / Rust "Features" groups alone — those are auto-generated.

---

## Files to reference while writing

- `praisonai-package/src/praisonai-train/praisonai_train/train/llm/trainer.py` — the whole file; specifically `TRAINING_METHODS`, `RESPONSE_MARKERS`, `TEMPLATE_TO_MARKERS`, `_MASK_LABELS`, `decide_masking`, `resolve_response_markers`, `TrainModel.resolve_method`, `TrainModel._require_columns`, `TrainModel.process_dataset`, `TrainModel.train_model`.
- `praisonai-package/src/praisonai-train/tests/unit/train/test_response_masking.py` — mirrors the exact resolution / routing behaviour that should be documented.
- `praisonai-package/src/praisonai-train/tests/unit/train/test_training_methods.py` — canonical examples of accepted / rejected dataset shapes per method.

## Acceptance checklist for the follow-up PR

- [ ] `docs/train.mdx` masking sections rewritten around the two-route model with updated error text and run-summary lines.
- [ ] New `docs/features/train-preference-tuning.mdx` created, following the AGENTS.md template.
- [ ] Config tables include `method`, `beta`, `max_prompt_length`, `desirable_weight`, `undesirable_weight` in `docs/train.mdx` **and** `docs/cli/train.mdx`.
- [ ] `docs/features/praisonai-train-package.mdx` "safety guarantees" bullet updated; DPO / ORPO / KTO one-liner added.
- [ ] `docs.json` adds the new page under Features (not Concepts, not auto-generated groups).
- [ ] Hero Mermaid uses the standard `#8B0000` / `#189AB4` / `#10B981` / `#F59E0B` / `#6366F1` palette with white text.
- [ ] All code examples are copy-paste runnable and use the friendly top-level imports where relevant.
- [ ] No edits to `docs/concepts/**` and no edits to `docs/sdk/reference/**`.
